programs/survival/waystones.sc

```diff
@@ -0,0 +1,310 @@
+// - pagination, currently the app will display 54 waypoints per player.
+
+
+global_settings = {
+    'dimensional_crossing'->false,
+    'enabled_dimensions'->'overworld',
+    'offering'->'ender_eye',
+    'clear_waypoints_on_death'->true,
+    'structure_material'->'copper',
+    'structure_size'->6
+};
+
+__config()->{
+    'command_permission'->'ops',
+    'scope'->'global',
+    'commands'->{
+        ''->'open_waystones_screen',
+        'setting offering <item>'->['setting_command','offering'],
+        'setting offering'->['setting_read','offering'],
+        'setting dimensionalCrossing <bool>'->['setting_command','dimensional_crossing'],
+        'setting dimensionalCrossing'->['setting_read','dimensional_crossing'],
+        'setting clearWaypointsOnDeath <bool>'->['setting_command','clear_waypoints_on_death'],
+        'setting clearWaypointsOnDeath'->['setting_read','clear_waypoints_on_death'],
+        'setting enabledDimensions <dimensions>'->['setting_command','enabled_dimensions'],
+        'setting enabledDimensions'->['setting_read','enabled_dimensions'],
+        'setting structureMaterial <dimensions>'->['setting_command','structure_material'],
+        'setting structureMaterial'->['setting_read','structure_material'],
+        'setting structureSize <int>'->['setting_command','structure_size'],
+        'setting structureSize'->['setting_read','structure_size'],
+    },
+    'arguments' -> {
+        'item'->{'type'->'term', 'suggest'->[
+            'ender_pearl', 'ender_eye', 'amethyst_shard', 'end_crystal', 'chorus_fruit', 'totem_of_undying']},
+        'setting'->{'type'->'term'},
+        'dimensions'->{'type'->'text', 'suggest'->['overworld', 'the_nether', 'the_end', 'overworld the_nether']}
+    }
+};
+
+
+// Keep track of all the waystones within the world.
+// [x,y,z] > name, dimension, icon
+global_waystones = {};
+// keep track of the waypoints each player has discovered.
+// uuid > [x,y,z](waystone pos) > [x,y,z](tp safe pos)
+global_waypoints = {};
+
+
+__on_start() -> (
+    _read_waystones();
+    file = read_file('waystone_settings', 'json');
+    if(file, global_settings = file);
+);
+__on_close() -> (
+    _write_waystones();
+    write_file('waystone_settings', 'json', global_settings);
+);
+
+_app_message(p, m) -> (
+    display_title(p, 'actionbar', format('wb '+m));
+);
+
+setting_read(key) -> (
+    print(player(), str('%s: %s', key, global_settings:key));
+);
+
+setting_command(val, key) -> (
+    global_settings:key = val;
+    print(player(), str('%s: %s', key, global_settings:key));
+);
+
+__on_player_places_block(p, item, hand, block)->(
+    if(item:0=='lodestone',
+        pos = pos(block);
+        floor = str(block(pos:0, pos:1 - 1, pos:2));
+        if(_has_required_structure(pos),
+            global_waystones:pos = {
+                'icon'->floor,
+                'dimension'->current_dimension(),
+                'name'->_get_name_from_nbt(item:2) || floor+' Waypoint' 
```

Comment:
Would look a lot better if you could get the block name, but I don't think scarpet can do that atm? Possible suggestion to scarpet tbh.